Core: Add `./` to start of hidden file & folder paths

IanVS opened the pull request:
Issue:

I'm trying to troubleshoot https://github.com/eirslett/storybook-builder-vite/issues/155, where I'm having trouble with stories that are defined inside of my `.storybook` folder. I found that storybook only adds a `./` to the start of a filename if it doesn't already find a single `.` at the start of the file, which inadvertently picks up hidden files & folders.  

## What I did

I've created a helper function to add the ./ when necessary, so it can be shared between the two places we know of where the same thing is done. I used @joshwooding's suggested snippet, because I couldn't think of anything better and I think it's pretty solid.  

## How to test

```
yarn test lib/core-common/src/utils/__tests__/paths.test.ts
yarn test lib/core-common/src/utils/__tests__/to-importFn.test.ts
yarn test lib/core-common/src/utils/__tests__/normalize-stories.test.ts
```

- [X] Is this testable with Jest or Chromatic screenshots?
- [ ] Does this need a new example in the kitchen sink apps?
- [ ] Does this need an update to the documentation?

If your answer is yes to any of these, please make sure to include it in your PR.

<!--

Everybody: Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `master` branch as part of the release process, so you shouldn't need to worry about this.

Maintainers: Please tag your pull request with at least one of the following:
`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

-->

IanVS committed:
Add tests for stories in config dir

IanVS committed:
Add configPath test for normalizeStoriesEntry

IanVS commented:
Feel free to close out this PR if it's not useful.  I've found the root cause of my problems, and submitted a PR to the vite builder: https://github.com/eirslett/storybook-builder-vite/pull/156, although I think the storybook behavior is a little bit brittle.

tmeasday commented on lib/core-common/src/utils/__tests__/normalize-stories.test.ts:
```diff
@@ -54,6 +54,25 @@ describe('normalizeStoriesEntry', () => {
     ]);
   });
 
+  it('story in config dir', () => {
+    const specifier = normalizeStoriesEntry('./file.stories.mdx', options);
+    expect(specifier).toMatchInlineSnapshot(`
+      {
+        "titlePrefix": "",
+        "directory": ".storybook",
+        "files": "file.stories.mdx",
+        "importPathMatcher": {}
+      }
+    `);
+
+    expect(specifier.importPathMatcher).toMatchPaths(['./.storybook/file.stories.mdx']);
+    expect(specifier.importPathMatcher).not.toMatchPaths([
+      '.storybook/file.stories.mdx',
+      './file.stories.mdx',
+      '../file.stories.mdx',
+    ]);
+  });
+
```
This seems great to me!

tmeasday commented on lib/core-common/src/utils/__tests__/to-importFn.test.ts:
```diff
@@ -160,6 +160,24 @@ const testCases: [string, string[], string[]][] = [
       '/Users/user/code/src/components/Icon.stories.js',
     ],
   ],
+  [
+    './Introduction.stories.tsx',
+    ['/Users/user/code/path/Introduction.stories.tsx'],
+    [
+      '/Users/user/code/Introduction.stories.tsx',
+      '/Users/user/code/src/Introduction.stories.tsx',
+      '/Users/user/code/src/Introduction.tsx',
+    ],
+  ],
+  [
+    'Introduction.stories.tsx',
+    ['/Users/user/code/path/Introduction.stories.tsx'],
+    [
+      '/Users/user/code/Introduction.stories.tsx',
+      '/Users/user/code/src/Introduction.stories.tsx',
+      '/Users/user/code/src/Introduction.tsx',
+    ],
+  ],
```
I guess these tests are less helpful because the working dir == the config dir here. Should we change that? It's sort of unrealistic

tmeasday reviewed (commented):
I like where this is going @IanVS. LMK if you need more help from me.

joshwooding commented:
Seems to me the bug is here https://github.com/storybookjs/storybook/blob/5da733d1711503a4c134b2d728eb467bce7641d1/lib/core-common/src/utils/normalize-stories.ts#L51-L53

I don't know if there's a nice way to write it but we need something to essentially do:
```ts
  if (!(directoryFromWorking.startsWith(`.${path.sep}`) || directoryFromWorking.startsWith(`..${path.sep}`))) {
    directoryFromWorking = `.${path.sep}${directoryFromWorking}`;
  }
```

IanVS commented:
Good catch, I think there are a few place spread around where this happens.  There's also https://github.com/storybookjs/storybook/blob/177dd455110cead4196cecae43e41738445b8252/lib/core-server/src/utils/StoryIndexGenerator.ts#L106, which I think is what determines what "fileName" is sent to the client for a given story id.

IanVS committed:
Properly normalize hidden files/folders

IanVS committed:
Separate configDir and workingDir for more realism

IanVS commented:
@tmeasday I've created a helper function to add the `./` when necessary, so it can be shared between the two places we know of where the same thing is done. I used @joshwooding's snippet, because I couldn't think of anything better and I think it's pretty solid.  

I've also split the `workingDir` and `configDir` in the tests as you suggested, but I have a failing case that I'm not sure whether it's a real problem or just due to the test setup.

The tests that I wrote for the `./` issue are now passing, at least.  I can't see the CircleCI results for the tests, and I can't seem to run them locally successfully.  LMK if there's something else I can do here.

IanVS commented:
`lib/store/src/autoTitle.test.ts` seems to be failing because windows path separators are used, but the test is running on linux, so `path.sep` isn't matching.  Any ideas what to do about that?

shilman commented:
@IanVS we use the `slash` library to normalize windows paths. if there's a windows test case that's failing, it's a legitimate failure and points to a missing `slash` in the code

IanVS commented:
> @IanVS we use the `slash` library to normalize windows paths. if there's a windows test case that's failing, it's a legitimate failure and points to a missing `slash` in the code

Thanks, but I think the problem is that the test is _faking a windows path_ by using `\\`, but since the code under test uses `path.sep`, it doesn't match (since the test is running on Linux).

shilman commented:
@IanVS unless those tests are new, in theory if it was passing before it means that the contract is that the code under test should be able to handle windows paths as input and did before this PR. our tests are a mess, so i'm not confident that theory matches practice in this case 🤷‍♂️

IanVS commented:
@shilman previously the code only looked for a `.` at the start of the path.  Now with the fix in this PR, it looks for a `.${path.sep}`.  We'd have to run the windows test on a windows box for it to be a good test.  Maybe to accommodate the test I should use a regex instead, and check for either a `/` or `\`.

shilman commented:
Works for me! 👍

joshwooding commented:
Using regex would cover all cases :) and you can consolidate the check into one regex

tmeasday commented on lib/core-common/src/utils/__tests__/paths.test.ts:
```diff
@@ -0,0 +1,34 @@
+  });
+
+  it('adds "./" to a hidden file', () => {
+    // .storybook/Comp.story.js
```
```suggestion
    // .Comp.story.js
```

tmeasday reviewed (approved):
Really great stuff @IanVS. I would like to test it on a windows box (unless you did already?) -- let me do that before we merge it.

IanVS committed:
Use regex to detect normalized relative paths

IanVS commented:
I've updated the check to use regex as discussed, and the windows-style path tests are now passing.  I also added some windows-style path tests to my new function as well. 

What is still failing, is a test in `to-importFn.test.ts`.  @tmeasday, can you tell if they are a legitimate failure, or just an artifact of how the test is being set up and run?  

Oh, and I don't have a windows box set up to run tests on, so if you do @tmeasday, and wouldn't mind running them, that would be much appreciated.  I think there's also a way to set up github actions to run on a windows vm...  I'll submit an experiment in a separate PR to see if that might be feasible.

shilman reviewed (changes requested):
Unit tests are still failing?

### toImportFn - webpackIncludeRegexp › matches only suitable paths - ../src/**/*.stories.tsx

```diff
    - Expected  - 1
    + Received  + 5

    - Array []
    + Array [
    +   "/Users/user/code/.storybook/src/Icon.stories.tsx",
    +   "/Users/user/code/.storybook/src/components/Icon.stories.tsx",
    +   "/Users/user/code/.storybook/src/components/Icon.stories/Icon.stories.tsx",
    + ]
```

IanVS commented:
Yes, it's unclear to me if those are legitimate failures (if so it I believe it would be an existing issue prior to this PR), or if it is due to the way the test itself is constructed.  I'm hoping @tmeasday has some insight.

tmeasday commented:
@IanVS I'm not sure exactly what's going on with these tests, but what's weird about this code is the `toImportFn` only matches the RHS of the path (that's because webpack passes the full absolute path into the regex, but we don't want to build absolute paths into the user's build).

So `'../src/**/*.stories.tsx'`, should match  `'/Users/user/code/src/Icon.stories.tsx'`, because in the original test the `configDir` was `'/path'`, so you'd end up with `'/src/**/*.stories.tsx'`, which does match the RHS of that filename.

Kind of a confusing/weird test though.

After your changes, the `configDir` is `'/Users/user/code/.storybook'` , so after resolution the directory would be `'/Users/user/code/src/**/stories.tsx'`, which I still would have thought matches?

I would have thought it was the *next* test `'../../src/**/*.stories.tsx'` that should have broken 🤔

tmeasday commented:
> After your changes, the configDir is the much more sane '/Users/user/code/.storybook' , so after resolution the directory would be '/Users/user/code/src/**/stories.tsx', which I still would have thought matches?

OK, retract that, the resolution is just relative to the `workingDir`, so in this case it would resolve to `./src/**/stories.tsx`

tmeasday commented:
Ahh, sorry I finally get it, sorry was a bit slow there. So yeah, the reason, is as I said:

> what's weird about this code is the toImportFn only matches the RHS of the path (that's because webpack passes the full absolute path into the regex, but we don't want to build absolute paths into the user's build).

So a matcher like `'./src/**/stories.tsx'` will end up matching `./.storybook/src/x/y/z.stories.tsx`, because we only check the RHS and don't build any paths into our match that aren't already in the SB (i.e. we don't build `workingDir` into the matcher).

IanVS commented:
My understanding must be flawed, but I would have thought that if the config dir is `/Users/user/code/.storybook`, and the pattern is relative to that directory, then `../src/**` wouldn't include anything inside `/Users/user/code/.storybook`, since they are sister directories.  The glob is going back up to the parent, and then down another path, right? I'm on discord for the next few minutes if it would be easier to chat there.  That said, if you think these are valid paths, I'm happy to either move them into the valid section, or delete them altogether since they're kind of wonky.  ;)

IanVS commented:
I talked with Tom, he's going to comment out those tests and leave a note on why they are not really possible to test (due to webpack limitations) just in case anyone wonders about it in the future.

tmeasday commented on lib/core-common/src/utils/__tests__/to-importFn.test.ts:
```diff
@@ -114,6 +117,10 @@ const testCases: [string, string[], string[]][] = [
     ],
     [
+      '/Users/user/code/.storybook/Icon.stories.tsx',
+      '/Users/user/code/.storybook/src/Icon.stories.tsx',
+      '/Users/user/code/.storybook/src/components/Icon.stories.tsx',
+      '/Users/user/code/.storybook/src/components/Icon.stories/Icon.stories.tsx',
```
```suggestion
      // Although it would make sense for these three files to fail to match the `importFn()`,
      // because we are limited to matching on the RHS of the path (from 'src' onwards, basically)
      // we cannot avoid matching things inside the config dir in such situations.
      // '/Users/user/code/.storybook/src/Icon.stories.tsx',
      // '/Users/user/code/.storybook/src/components/Icon.stories.tsx',
      // '/Users/user/code/.storybook/src/components/Icon.stories/Icon.stories.tsx',
```

tmeasday committed:
Comment out tests w/ comment explaining why
